### ISSUE TITLE
Wait for ng_binding

### ISSUE THREAD

**Issue by ehannes** [kikonen/capybara-ng#16]
I use Capybara with the webkit driver to test AngularJS on my application. When I fill in a form and pressing submit, it seems as if Capybara does not wait for Angular to bind the value to the corresponding model. Does this gem provide a method for this kind of waiting? Can I tell Capybara to wait until Angular has done all bindings? Or can I tell Capybara to wait until a specific binding is done?

**Comment by ehannes** [comment]
I think it would be more elegant to find a way to wait for Angular to do its bindings, but if its not possible I guess I could use the `ng_model().set()` method from this gem to set the value. I don't know if I'm doing it wrong, but the following fails:
```
ng_model('model.attribute').set('false')
expect(has_ng_model?('model.attribute')).to eq false

      Failure/Error: expect(has_ng_model?('model.attribute')).to eq false
     
       expected: false
            got: true
     
       (compared using ==)
```

**Comment by ehannes** [comment]
I managed to get my test passing by setting the attribute of ngModel explicitly with `ng_model('model.attribute').set('value')`, but I think it would be more elegant to find a way to wait for Angular to do its bindings.

**Comment by kikonen** [comment]
There is some base logic in this gem, which is supposed to be waiting if angularjs is ready (lib/angular/waiter.rb). But now when I'm looking into logic, it feels me that this logic is not doing desired thing (i.e. waiting angularjs to become "ready" in every moment).

**Comment by kikonen** [comment]
okey, miss recalled logic...

ng_wait logic is trying to wait for stable point in angularjs via "angular.getTestability.whenStable()", and by default this is done for every "ng_xxx" operation.

Perhaps, logic should add "digestable" action in whenStable -hook, and set "ready" flag only after it is processed. I presume something like that should be doing what you are thinking.

**Comment by kikonen** [comment]
Not 100% sure if it will help you, but I changed "ng_wait" semantics in two ways

1) Angular.wait_time determines wait time for timeout in waiting (this in case capybara normal wait time is tuned to be intolerably small)

2) ng_wait does internally $timeout trick to trigger $digest loop, which should ensure that all pending model updates are done.

**Comment by ehannes** [comment]
@kikonen:
Thanks for your time! I tried to update to latest version (0.2.7), but I still have the same problem.

I have something like:
```
background do
  visit some_page

  uncheck 'some_checkbox'
  click_button 'save'
  expect(page).to have_content('Some model is saved!')

  visit some_page # Reload page
end

scenario do
  expect(page).to have_unchecked_field('some_checkbox')
end
```

I tried to add `ng_wait` after I uncheck `some_checkbox`, but that didn't help. 

As I said before, adding `ng_model('some_angular_model.some_checkbox').set(false)` does work, but it's ugly.

**Comment by kikonen** [comment]
At least there should be "ng_wait"  before
```
expect(page).to have_content('Some model is saved!')
```

**Comment by kikonen** [comment]
After visit_page you want to say ng_wait to wait angular to br ready so that it has change to react to DOM changes and events

**Comment by ehannes** [comment]
Yes, "save" is an async Angular operation. But I have no problem waiting for `'Some model is saved!'` since Capybara handles that one perfectly. The problem is that the checkbox is not checked for some reason before the save button is pressed.

**Comment by ehannes** [comment]
I actually managed to get this particular example working now. I had omitted some (many) details to describe my problem. I have a select statement right before the checkbox, and it seams as if the select box may have blocked the checkbox. Now I simulate blur after I have selected an option and as I said, this particular example is now working.

```
background do
  visit some_page

  select 'some_option', from: 'some_select_input'
  find('body').click # simulate blur
  uncheck 'some_checkbox'
  click_button 'save'
  expect(page).to have_content('Some model is saved!')

  visit some_page # Reload page
end

scenario do
  expect(page).to have_select('some_select_input', selected: 'some_option')
  expect(page).to have_unchecked_field('some_checkbox')
end
```

I'll keep testing your gem and see if it'll help me with the rest of the problems I've experienced. Thanks for your time!

**Comment by ehannes** [comment]
Still struggling with this problem... You wrote that
> ng_wait does internally $timeout trick to trigger $digest loop, which should ensure that all pending model updates are done.

Is there any way to ensure the page is updated, that Angular is ready to react?

A lot of my problems seems to be related with radio buttons. I choose between two values and this choice determines what fields are being showed. Sometimes, when the test selects a certain radio button, the ng_model is not updated to the correct value corresponding to the radio button and the correct fields are not displayed. Can this be because the page is not fully loaded when the test chooses a value for the radio button so that Angular cannot react? Is there any way to make sure that Angular is ready? Is this what `ng_wait` does? I have tried it, as you suggested above after visiting a page for instance, but the test still fails randomly.

**Comment by kikonen** [comment]
Yes, ng_wait tries to wait that angularjs is ready. But does readiness of your app depend from asyn xhr calls in startup? I would suggest adding redundant console logging around logic dealing with radio buttons, it might reveal something.

**Comment by kikonen** [comment]
ng_eval("some expr") could be also useful to track down the issue.

I.e. via https://www.ng-book.com/p/Debugging-AngularJS/
